go/install/update.go

```diff
@@ -0,0 +1,377 @@
+	u.G().Log.Info("Checking for update, current version is %s", u.config.Version)
+
+	currentSemVersion, err := semver.Make(u.config.Version)
+	if err != nil {
+		return
+	}
+
+	update, err = u.source.FindUpdate(u.config)
+	if err != nil || update == nil {
+		return
+	}
+
+	u.G().Log.Info("Found update with version: %s", update.Version)
+	updateSemVersion, err := semver.Make(update.Version)
+	if err != nil {
+		return
+	}
+	if updateSemVersion.EQ(currentSemVersion) {
+		// Versions are the same, we are up to date
+		return
+	} else if updateSemVersion.LT(currentSemVersion) {
+		err = fmt.Errorf("Update is older version: %s < %s", updateSemVersion, currentSemVersion)
+		return
+	}
+
+	return
+}
+
+func computeEtag(path string) (string, error) {
+	var result []byte
+	file, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(result)), nil
+}
+
+func (u *Updater) pathForFilename(filename string) string {
+	return path.Join(os.TempDir(), "KeybaseUpdates", filename)
+}
+
+func isFileScheme(url string) (string, bool) {
+	scheme := "file://"
+	if strings.HasPrefix(url, scheme) {
+		fpath := url[len(scheme):]
+		return fpath, true
+	}
+
+	return "", false
+}
+
+func (u *Updater) downloadAsset(asset keybase1.Asset) (fpath string, cached bool, err error) {
+	url := asset.DownloadURL
+	if lpath, b := isFileScheme(url); b {
+		u.G().Log.Info("Using local path: %s", lpath)
+		fpath = lpath
+		return
+	}
+
+	filename := asset.Name
+	fpath = u.pathForFilename(filename)
+	err = libkb.MakeParentDirs(fpath)
+	if err != nil {
+		return
+	}
+	etag := ""
+	if _, err = os.Stat(fpath); err == nil {
+		etag, err = computeEtag(fpath)
+		if err != nil {
+			return
+		}
+	}
+
+	req, _ := http.NewRequest("GET", url, nil)
+	u.G().Log.Info("Using etag: %s", etag)
+	if etag != "" {
+		req.Header.Set("If-None-Match", etag)
+	}
+	client := &http.Client{}
+	u.G().Log.Info("Request %#v", url)
+	resp, err := client.Do(req)
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotModified {
+		cached = true
+		return
+	}
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("Responded with %s", resp.Status)
+		return
+	}
+
+	savePath := fmt.Sprintf("%s.download", fpath)
+	file, err := os.OpenFile(savePath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, libkb.PermFile)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+	n, err := io.Copy(file, resp.Body)
+	u.G().Log.Info("Saving %d bytes to %s", n, fpath)
+	if err != nil {
+		return
+	}
+
+	if _, err = os.Stat(fpath); err == nil {
+		u.G().Log.Info("Removing existing bad download: %s", fpath)
+		err = os.Remove(fpath)
+		if err != nil {
+			return
+		}
+	}
+	u.G().Log.Info("Moving %s to %s", path.Base(savePath), path.Base(fpath))
+	err = os.Rename(savePath, fpath)
+	return
+}
+
+func (u *Updater) unpack(filename string) (string, error) {
```

Comment:
What's our story going to be for signing and verifying new downloads?  All within the DMG?  We have the opportunity to check here too, and maybe in a way that doesn't depend on the Root CAs or Apple.